components/datetime/src/options/components.rs

```diff
@@ -1,19 +1,26 @@
 // This file is part of ICU4X. For terms of use, please see the file
 // called LICENSE at the top level of the ICU4X source tree
 // (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
-//! Components is a model of encoding information on how to format date and time by specifying a list of components
-//! the user wants to be visible in the formatted string and how each field should be displayed.
+
+//! (unimplemented) `Components` is an options type available to [`DateTimeFormatOptions`](crate::DateTimeFormatOptions).
+//! It is a way of encoding information on how to format date and time by specifying an un-ordered
+//! "bag" of components that the user wants to be visible in the formatted string, plus defining how each field should
+//! be displayed. The ordering and output of the final string is generated by the
+//! [`DateTimeFormat`](crate::DateTimeFormat) using the configuration of the [`components::Bag`](crate::options::components::Bag),
+//! combined with the available localization data.
 //!
-//! This model closely corresponds to `ECMA402` API and allows for high level of customization compared to `Style` model.
+//! Additionally, the [`components::Bag`](crate::options::components::Bag) can contain an optional bag of `Preferences` which represents
+//! user preferred adjustments that can be applied onto the pattern right before formatting.
 //!
-//! Additionally, the bag contains an optional set of `Preferences` which represent user preferred adjustments
-//! that can be applied onto the pattern right before formatting.
+//! This `Components` model closely corresponds to `ECMA402` API and allows for high level of customization
+//! compared to `Style` model.
 //!
 //! # Pattern Selection
 //!
-//! It is important to understand that the components bag is a human-friendly way to describe a skeleton, not a pattern.
-//! That means that the components and their styles provided by the user will be matched against available patterns for
-//! a given locale and the closest available pattern will be used for formatting.
+//! It is important to understand that the components bag is a human-friendly way to describe a skeleton of how to
```

Comment:
I still don't really understand what `skeleton` means here still. It's the first use of this term here without a definition.

Reply:
Great catch! It's a bit documented here: https://unicode.org/reports/tr35/tr35-dates.html#availableFormats_appendItems but I feel like Unicode generally underdocuments skeletons.

If I were to describe it, I'd go for something like:

```
Skeletons are locale-agnostic identifiers containing field information, and in a canonical order. Examples are "yMMMM" for year + full month, or "MMMd" for abbreviated month + day.
```

Reply:
> canonical order

It seems like skeleton is the incorrect term here then, as the components doesn't provide an order for these terms, it's an unordered collection.

Reply:
The canonicality (is it a word?) of the order is only necessary because skeleton is a string. When representing a bag like:

```
{
  "year": "long",
  "month": "2-digit",
  "day": "2-digit"
}
```

as a bag, I can set any order, and the output will be the same. But when encoding it as a skeleton, we need to make sure that we don't end up with `yyyyMMDD` being different from `MMyyyyDD` and `DDMMyyyy`.
So when encoding skeletons as strings we use "canonical" order of year->month->day, and in result every skeleton will be a string encoded with fields in that order.

So, bag is a representation of a skeleton. It is unordered, but its hashing/stringification will be ordered.

Does that make sense?

Reply:
I use the term "skeleton" mostly interchangeably with what Zibi is defining as "components".  A skeleton can be represented either as a string (UTS 35 style) or as an options bag.  With this mental model, we could consider the Components class as having a FromStr that parses the UTS 35-style skeleton string.

Reply:
> Skeletons are locale-agnostic identifiers containing field information in a canonical order.

The order does not matter.
"yMMMd" and "MMMyd" is the same skeleton, with the same result.

You can also "sprinkle" other characters, and they will be ignored:
"MMMM d, y" is a valid skeleton (and is the same as "yMMMMd")
But I don't like it.
I prefer ISO order and no skeleton, not a pattern.